### ISSUE TITLE
/EDIT/GET errors with "Error: send EMSGSIZE"

### ISSUE THREAD

**Issue by lhkev** [jean-emmanuel/open-stage-control#575]
- Open Stage Control version (or commit if built from sources)
version 0.48.5
- Operating system
Mojave 10.14.4
- Open Stage Control options
 --load edit-get-error.json \
 --debug true \
 --theme orange

- Client type (built-in or external chromium/chrome browser) and version
Safari 
- Session file (in ZIP file)
edit-get-error.json
[OSC-Bugs-Demos-2.zip](https://github.com/jean-emmanuel/open-stage-control/files/3906700/OSC-Bugs-Demos-2.zip)
- Custom module
none
- Server logs (from the terminal or the launcher's console)
{ Error: send EMSGSIZE 127.0.0.1:8080
    at Object.exports._errnoException (util.js:1024:11)
    at exports._exceptionWithHostPort (util.js:1047:20)
    at SendWrap.afterSend [as oncomplete] (dgram.js:474:11)
  code: 'EMSGSIZE',
  errno: 'EMSGSIZE',
  syscall: 'send',
  address: '127.0.0.1',
  port: 8080 }

Attached is a ZIP file containing a session file: edit-get-error.json and a command file to run the session.

The session contains two tabs. The first tab contains one button which sends the OSC remote control message:
`/EDIT/GET`
with preargs:

`[
  "127.0.0.1:8080",
  "templates"
]`

'templates' is the id of the second tab. 

The second tab contains several widgets. (In the project I am creating, this tab contains templates for widgets that will be created programatically in a custom module. The custom module retrieves the templates using /EDIT/GET, stores them as objects, and then uses these to create new instances of the widgets as required).

Pressing the push button on the first tab will show the issue in the log.

When the number of widgets in the second tab causes the length of the JSON to exceed a certain value, the OSC send fails with EMSGSIZE error.

In the log you will see the constructed JSON, followed by the error when attempting to send the JSON.

[OSC-Bugs-Demos-2.zip](https://github.com/jean-emmanuel/open-stage-control/files/3906700/OSC-Bugs-Demos-2.zip)

Thank you for your kind attention.

log from bug demo:

`App available at http://127.0.0.1:8080 & http://192.168.248.128:8080
OSC sent:  { address: '/EDIT/GET',
  args: 
   [ { type: 's', value: '127.0.0.1:8080' },
     { type: 's', value: 'templates' } ] } To: 127.0.0.1:8080
OSC received:  { address: '/EDIT/GET',
  args: [ '127.0.0.1:8080', 'templates' ] } From: 127.0.0.1:8080
OSC sent:  { address: '/EDIT/GET',
  args: 
   [ { type: 's', value: 'templates' },
     { type: 's',
       value: '{"type":"tab","id":"templates","linkId":"","label":"auto","color":"auto","css":".widget {text-transform:none;}","default":"","value":"","precision":2,"address":"/tab_4","preArgs":"","target":"","bypass":false,"variables":"@{parent.variables}","widgets":[{"type":"panel","widgets":[{"type":"text","top":0,"left":0,"id":"template_cuetype","linkId":"","width":120,"height":40,"label":false,"color":"auto","css":"","vertical":false,"wrap":false,"align":"","default":"","value":"","address":"/template_cuetype","preArgs":""},{"type":"text","top":0,"left":130,"id":"template_cuenumber","linkId":"","width":170,"height":40,"label":false,"color":"auto","css":"","vertical":false,"wrap":false,"align":"","default":"","value":"","address":"/template_cuenumber","preArgs":""},{"type":"text","top":0,"left":310,"id":"template_cuename","linkId":"","width":500,"height":40,"label":false,"color":"auto","css":"","vertical":false,"wrap":true,"align":"left","default":"","value":"","address":"/template_cuename","preArgs":""}],"label":false,"top":30,"left":10,"id":"template_cuepanel","linkId":"","width":840,"height":40,"color":"lightblue","css":"","scroll":true,"border":true,"default":"","value":"","precision":2,"address":"/template_cuepanel","preArgs":"","target":"","bypass":false,"variables":"@{parent.variables}","tabs":[]},{"type":"panel","widgets":[{"type":"strip","widgets":[{"type":"knob","top":40,"left":380,"id":"template_knob_compact0","linkId":"","width":90,"height":80,"label":false,"color":"auto","css":"","range":{"min":-60,"80%":0,"73%":-5,"65%":-10,"53%":-20,"39%":-30,"26%":-40,"12%":-50,"max":12},"input":true,"compact":true,"pips":true,"snap":false,"spring":false,"logScale":false,"sensitivity":0.2,"default":"","value":"","precision":2,"address":"/cue_id/@{parent.variables.qlabid}/sliderLevel","preArgs":0,"target":"","bypass":false,"touchAddress":"","dashed":true,"angle":300,"doubleTap":false,"steps":73,"unit":"","origin":"auto"},{"type":"knob","top":30,"left":330,"id":"template_knob_compact1","linkId":"","width":90,"height":80,"label":false,"color":"auto","css":"","range":{"min":-60,"80%":0,"73%":-5,"65%":-10,"53%":-20,"39%":-30,"26%":-40,"12%":-50,"max":12},"input":true,"compact":true,"pips":true,"snap":false,"spring":false,"logScale":false,"sensitivity":0.2,"default":"","value":"","precision":2,"address":"/cue_id/@{parent.variables.qlabid}/sliderLevel","preArgs":1,"target":"","bypass":false,"touchAddress":"","dashed":true,"angle":300,"doubleTap":false,"steps":73,"unit":"","origin":"auto"},{"type":"knob","top":40,"left":320,"id":"template_knob_compact2","linkId":"","width":90,"height":80,"label":false,"color":"auto","css":"","range":{"min":-60,"80%":0,"73%":-5,"65%":-10,"53%":-20,"39%":-30,"26%":-40,"12%":-50,"max":12},"input":true,"compact":true,"pips":true,"snap":false,"spring":false,"logScale":false,"sensitivity":0.2,"default":"","value":"","precision":2,"address":"/cue_id/@{parent.variables.qlabid}/sliderLevel","preArgs":2,"target":"","bypass":false,"touchAddress":"","dashed":true,"angle":300,"doubleTap":false,"steps":73,"unit":"","origin":"auto"},{"type":"knob","top":30,"left":280,"id":"template_knob_compact3","linkId":"","width":90,"height":80,"label":false,"color":"auto","css":"","range":{"min":-60,"80%":0,"73%":-5,"65%":-10,"53%":-20,"39%":-30,"26%":-40,"12%":-50,"max":12},"input":true,"compact":true,"pips":true,"snap":false,"spring":false,"logScale":false,"sensitivity":0.2,"default":"","value":"","precision":2,"address":"/cue_id/@{parent.variables.qlabid}/sliderLevel","preArgs":3,"target":"","bypass":false,"touchAddress":"","dashed":true,"angle":300,"doubleTap":false,"steps":73,"unit":"","origin":"auto"},{"type":"knob","top":40,"left":350,"id":"template_knob_compact4","linkId":"","width":90,"height":80,"label":false,"color":"auto","css":"","range":{"min":-60,"80%":0,"73%":-5,"65%":-10,"53%":-20,"39%":-30,"26%":-40,"12%":-50,"max":12},"input":true,"compact":true,"pips":true,"snap":false,"spring":false,"logScale":false,"sensitivity":0.2,"default":"","value":"","precision":2,"address":"/cue_id/@{parent.variables.qlabid}/sliderLevel","preArgs":4,"target":"","bypass":false,"touchAddress":"","dashed":true,"angle":300,"doubleTap":false,"steps":73,"unit":"","origin":"auto"},{"type":"knob","top":50,"left":450,"id":"template_knob_compact5","linkId":"","width":90,"height":80,"label":false,"color":"auto","css":"","range":{"min":-60,"80%":0,"73%":-5,"65%":-10,"53%":-20,"39%":-30,"26%":-40,"12%":-50,"max":12},"input":true,"compact":true,"pips":true,"snap":false,"spring":false,"logScale":false,"sensitivity":0.2,"default":"","value":"","precision":2,"address":"/cue_id/@{parent.variables.qlabid}/sliderLevel","preArgs":5,"target":"","bypass":false,"touchAddress":"","dashed":true,"angle":300,"doubleTap":false,"steps":73,"unit":"","origin":"auto"},{"type":"knob","top":40,"left":510,"id":"template_knob_compact6","linkId":"","width":90,"height":80,"label":false,"color":"auto","css":"","range":{"min":-60,"80%":0,"73%":-5,"65%":-10,"53%":-20,"39%":-30,"26%":-40,"12%":-50,"max":12},"input":true,"compact":true,"pips":true,"snap":false,"spring":false,"logScale":false,"sensitivity":0.2,"default":"","value":"","precision":2,"address":"/cue_id/@{parent.variables.qlabid}/sliderLevel","preArgs":6,"target":"","bypass":false,"touchAddress":"","dashed":true,"angle":300,"doubleTap":false,"steps":73,"unit":"","origin":"auto"},{"type":"knob","top":40,"left":580,"id":"template_knob_compact7","linkId":"","width":90,"height":80,"label":false,"color":"auto","css":"","range":{"min":-60,"80%":0,"73%":-5,"65%":-10,"53%":-20,"39%":-30,"26%":-40,"12%":-50,"max":12},"input":true,"compact":true,"pips":true,"snap":false,"spring":false,"logScale":false,"sensitivity":0.2,"default":"","value":"","precision":2,"address":"/cue_id/@{parent.variables.qlabid}/sliderLevel","preArgs":7,"target":"","bypass":false,"touchAddress":"","dashed":true,"angle":300,"doubleTap":false,"steps":73,"unit":"","origin":"auto"},{"type":"knob","top":30,"left":750,"id":"template_knob_compact8","linkId":"","width":90,"height":80,"label":false,"color":"auto","css":"","range":{"min":-60,"80%":0,"73%":-5,"65%":-10,"53%":-20,"39%":-30,"26%":-40,"12%":-50,"max":12},"input":true,"compact":true,"pips":true,"snap":false,"spring":false,"logScale":false,"sensitivity":0.2,"default":"","value":"","precision":2,"address":"/cue_id/@{parent.variables.qlabid}/sliderLevel","preArgs":8,"target":"","bypass":false,"touchAddress":"","dashed":true,"angle":300,"doubleTap":false,"steps":73,"unit":"","origin":"auto"}],"label":false,"top":30,"left":0,"id":"template_knobstrip","linkId":"","width":830,"height":80,"color":"auto","css":":host > .panel {\\n\\n    background-color:darkslateblue;\\n\\n}","scroll":true,"horizontal":true,"stretch":false,"border":true,"spacing":0,"default":"","value":"","precision":2,"address":"/strip_3","preArgs":"","target":"","bypass":false,"variables":"@{parent.variables}","tabs":[]},{"type":"strip","widgets":[{"type":"text","label":false,"top":0,"left":0,"id":"cueout0","linkId":"","width":90,"height":30,"color":"auto","css":"","default":"","value":"Master","address":"/cueout0","preArgs":"","vertical":false,"wrap":false,"align":""},{"type":"text","top":0,"left":90,"id":"cueout1","linkId":"","width":90,"height":30,"label":false,"color":"auto","css":"","vertical":false,"wrap":false,"align":"","default":"","value":1,"address":"/cueout1","preArgs":""},{"type":"text","top":0,"left":180,"id":"cueout2","linkId":"","width":90,"height":30,"label":false,"color":"auto","css":"","vertical":false,"wrap":false,"align":"","default":"","value":2,"address":"/cueout2","preArgs":""},{"type":"text","top":0,"left":270,"id":"cueout3","linkId":"","width":90,"height":30,"label":false,"color":"auto","css":"","vertical":false,"wrap":false,"align":"","default":"","value":3,"address":"/cueout3","preArgs":""},{"type":"text","top":0,"left":360,"id":"cueout4","linkId":"","width":90,"height":30,"label":false,"color":"auto","css":"","vertical":false,"wrap":false,"align":"","default":"","value":4,"address":"/cueout4","preArgs":""},{"type":"text","top":0,"left":450,"id":"cueout1","linkId":"","width":90,"height":30,"label":false,"color":"auto","css":"","vertical":false,"wrap":false,"align":"","default":"","value":5,"address":"/cueout1","preArgs":""},{"type":"text","top":0,"left":540,"id":"cueout5","linkId":"","width":90,"height":30,"label":false,"color":"auto","css":"","vertical":false,"wrap":false,"align":"","default":"","value":6,"address":"/cueout5","preArgs":""},{"type":"text","top":0,"left":630,"id":"cueout6","linkId":"","width":90,"height":30,"label":false,"color":"auto","css":"","vertical":false,"wrap":false,"align":"","default":"","value":7,"address":"/cueout6","preArgs":""},{"type":"text","top":0,"left":720,"id":"cueout7","linkId":"","width":90,"height":30,"label":false,"color":"auto","css":"","vertical":false,"wrap":false,"align":"","default":"","value":8,"address":"/cueout7","preArgs":""}],"label":false,"top":0,"left":0,"id":"strip_5","linkId":"","width":830,"height":30,"color":"auto","css":"","scroll":true,"horizontal":true,"stretch":false,"border":true,"spacing":0,"default":"","value":"","precision":2,"address":"/strip_5","preArgs":"","target":"","bypass":false,"variables":"@{parent.variables}","tabs":[]}],"label":false,"top":90,"left":10,"id":"template_audiopanel","linkId":"","width":840,"height":120,"color":"lightgreen","css":"","scroll":true,"border":true,"default":"","value":"","precision":2,"address":"/template_audiopanel","preArgs":"","target":"","bypass":false,"variables":"@{parent.variables}","tabs":[]},{"type":"matrix","top":230,"left":10,"id":"template_fadermatrix","linkId":"","width":720,"height":330,"label":"auto","color":"auto","css":"","widgetType":"fader","matrix":[9,1],"start":0,"spacing":0,"traversing":true,"border":true,"props":"{\\n  \\"range\\": { \\"min\\": -60, \\"80%\\": 0, \\"73%\\": -5, \\"65%\\": -10, \\"53%\\": -20, \\"39%\\": -30, \\"26%\\": -40, \\"12%\\": -50, \\"max\\": 12 },\\n  \\"address\\": \\"/cue_id/@{this.id}/sliderLevel\\",\\n  \\"preArgs\\": \\"#{$}\\",\\n  \\"meter\\": true,\\n  \\"meterAddress\\": \\"/reply/cue_id/@{this.id}/liveAverageLevel/#{$}/meter\\",\\n  \\"label\\": \\"#{[ \\"Master\\", 1, 2, 3, 4, 5, 6, 7, 8 ][$]}\\"\\n}\\n","default":"","value":"","precision":2,"address":null,"preArgs":"","target":"","bypass":false}],"tabs":[],"scroll":true}' } ] } To: 127.0.0.1:8080
{ Error: send EMSGSIZE 127.0.0.1:8080
    at Object.exports._errnoException (util.js:1024:11)
    at exports._exceptionWithHostPort (util.js:1047:20)
    at SendWrap.afterSend [as oncomplete] (dgram.js:474:11)
  code: 'EMSGSIZE',
  errno: 'EMSGSIZE',
  syscall: 'send',
  address: '127.0.0.1',
  port: 8080 }

`

**Comment by jean-emmanuel** [comment]
Hi, the maximum packet size depends on the system's udp/tcp buffer size, you'll need to increase it to send bigger packets. The procedure depends on your os, it's pretty straightfoward on linux but I don't know about os x.
That being said, if you only need to process the response packet in your custom module, you can catch it in `oscOutFilter` and discard the udp/tcp message that's irrelevant.

```js
module.exports = {

  oscOutFilter: function(data) {

    if (data.address == '/EDIT/GET' && data.args[0].value == 'templates') {

      // process  here

      return // empty return to bypass message

    }

    return data

  }

}

**Comment by lhkev** [comment]
Hi and thanks for the suggestion to handle the response in the oscOutFilter function. That way the response does not pass over the loopback interface which is both logical and an efficiency saving, as well as avoiding any limitations on the interface.

I remained curious as to the cause of the problem on OSX. 

First I moved my development project onto Windows with Open Stage Control running on windows, and accessing Qlab on the Mac (Qlab only runs on Mac). Everything ran without issue.

But, on the Mac, OSC was receiving larger datagrams from qlab without issue, and sending larger datagrams to qlab without issue. So how come these datagrams passed across the loopback interface successfully, and yet the remote control response - entirely within OSC but still over the loopback interface - failed with the size error?

My own knowledge of OSX is limited, I only use MAC for qlab, and that is because one of my clients uses qlab, giving me no other option. It took a bit of delving... but I have found the cause.

The communications between OSC and Qlab which worked with large datagrams is using TCP/IP. However, it seems that the /EDIT/GET response was being sent by OSC using UDP, and there it hit an arbitrary limitation in OSX for UDP packet size.

In OSX, Apple have placed an arbitrary limit of 9216 bytes on UDP packets. It is possible to increase this value. The following command will increase the value - until the machine is rebooted:

`sudo sysctl -w net.inet.udp.maxdgram=65535`

(It is possible to make the change permanent using the /etc/sysctl.conf file).

When I made this change, the /EDIT/GET response passed without error.

If OSC had used TCP/IP then the /EDIT/GET transmission would have worked.

From this I assume that sending to a target is by default using UDP. I wonder if it would be possible to have an option to use TCP rather than UDP when sending to targets? It would certainly help anyone using OSC on a Mac. 

Again, thanks for your advice to use oscOutFilter.
